cli/flox-watchdog/src/process/macos.rs

```diff
@@ -0,0 +1,245 @@
+    /// that spawned this watchdog.
+    fn init_kqueue(&mut self) -> Result<(), Error> {
+        let mut kw = kqueue::Watcher::new().context("failed to create kqueue watcher")?;
+        kw.add_pid(
+            self.original_pid.into(),
+            kqueue::EventFilter::EVFILT_PROC,
+            kqueue::FilterFlag::NOTE_EXIT,
+        )?;
+        self.pids_watching.insert(self.original_pid);
+        kw.watch().context("failed to initialize kqueue watcher")?;
+        self.kqueue_watcher = Some(kw);
+        Ok(())
+    }
+
+    /// Add a PID to the kqueue watcher to get termination notifications.
+    fn add_pid(&mut self, pid: ActivationPid) -> Result<(), Error> {
+        if let Some(ref mut kw) = self.kqueue_watcher {
+            kw.add_pid(
+                pid.into(),
+                kqueue::EventFilter::EVFILT_PROC,
+                kqueue::FilterFlag::NOTE_EXIT,
+            )
+            .with_context(|| format!("failed to add PID {pid} to watch list"))
+        } else {
+            Err(anyhow!("tried to watch PID without initializing kqueue"))
+        }
+    }
+
+    /// Prunes any activation PIDs that have terminated
+    fn prune_terminations(&mut self) -> Result<(), Error> {
+        if let Some(ref mut kw) = self.kqueue_watcher {
+            while let Some(event) = kw.poll(None) {
+                if let Ident::Pid(pid) = event.ident {
+                    if !self.pids_watching.remove(&pid.into()) {
+                        warn!("received notification for PID that wasn't watched");
+                    }
+                }
+            }
+        } else {
+            bail!("tried to prune activations without initializing kqueue");
+        }
+        Ok(())
+    }
+}
+
+impl Watcher for MacOsWatcher {
+    fn wait_for_termination(&mut self) -> Result<WaitResult, Error> {
+        self.init_kqueue()?;
+        loop {
+            self.update_watchlist()?;
+            if self.should_clean_up()? {
+                return Ok(WaitResult::CleanUp);
+            }
+            if self
+                .should_terminate_flag
+                .load(std::sync::atomic::Ordering::SeqCst)
+            {
+                return Ok(WaitResult::Terminate);
+            }
+            if self
+                .should_clean_up_flag
+                .load(std::sync::atomic::Ordering::SeqCst)
+            {
+                return Ok(WaitResult::CleanUp);
+            }
+            std::thread::sleep(WATCHER_SLEEP_INTERVAL);
+        }
+    }
+
+    /// Update the list of PIDs that are currently being watched.
+    fn update_watchlist(&mut self) -> Result<(), Error> {
+        let all_registered_pids = activation_pids(&self.reg_path, &self.hash)?;
+        self.prune_terminations()?;
+        let to_add = self
+            .pids_watching
+            .difference(&all_registered_pids)
+            .cloned()
+            .collect::<Vec<_>>();
+        for pid in to_add {
+            if pid.is_running() {
```

Comment:
blocking: won't this add zombie watchdogs back into `pids_watching`?

Seems like we need to do one of the following:
1. Store a list of PIDs kqueue has told us exited. If a new activation is started with the same PID as a prior activation, we ignore it (bummer)
2. Don't use kqueue at all, and treat `(registry) minus (whatever's dead)` as the source of truth. If there's a zombie watchdog, we don't detect that (bummer)
3. Use kqueue to update the registry. There's still a race: activate 1 adds itself to the registry and exits immediately, activate 2 gets started with the same PID as activate 1 and adds itself to the registry, we poll kqueue and see activate 1 has exited so we remove PID from the registry thinking it's for activate 1, but it's now for activate 2 (bummer)

Bummer 3 seems to be the least likely.
I think doing 2 is probably the easiest because it's so symmetrical to what we'll be doing on Linux. I think that would mean we could use almost all the same code for watching on both platforms, and the only platform specific functionality would be `is_running()`. So without thinking through it more that would be my vote

Reply:
1. This doesn't really address the problem I had in mind, which is asking kqueue to watch a PID that's already terminated since that will cause a failure. I think this option is a non-starter.
2. Possible, but a real bummer. `kill -0` was the cause of a some `expect` test failures in the past, so we would need to rework those tests.
3. Right, this isn't very likely, but still a bummer.

The most symmetrical option to what we do on Linux would be using `libproc`, which is technically a macOS private API, but I think it's the library used by macOS `/bin/ps`. There are Rust bindings we could use. This would _really_ mean that the `Watcher` is a single struct with a `#[cfg(...)]` to switch on a small method for reading `/proc` or calling `libproc`.

The online docs for the Rust `libproc` crate are missing most of the relevant information because the crate is actually cross platform, but it looks like docs were only generated for Linux. Helpful. I think you would call `pidinfo::<BSDInfo>(pid, 0)`, which gives you a Rust version of `struct proc_bsdinfo`, which contains a [status field](https://github.com/apple-oss-distributions/xnu/blob/8d741a5de7ff4191bf97d57b9f54c2f6d4a15585/bsd/sys/proc_info.h#L61). This would mean that we could avoid the issues with `kill -0` and unify the implementations pretty easily (and also remove most of the platform dependent code since `libproc` is cross-platform).

Reply:
We decided to go with calling `/bin/ps` because getting the Nix build to work linking against the `libproc.h` headers turned out to be complicated.